examples/fairmotion_interface.py

```diff
@@ -546,9 +566,298 @@ def belongs_to(self, obj_id: int) -> bool:
 
         return False
 
+    def setup_pathfollower(self, path: habitat_sim.ShortestPath()):
+        """
+        Prepare the pathfollowing character and any data needed to execute the update function.
+        """
+        self.hide_model()
+        self.activity = Activity.PATH_FOLLOW_SEQ
+
+        self.path_points = path.points
+        self.path_time: float = 0.0  # tracks time along path
+        self.path_motion = Move.walk_to_walk.motion
+
+        # get path length
+        i, j, summ = 0, 0, 0.0
+        while i < len(self.path_points):
+            summ += (mn.Vector3(self.path_points[i] - self.path_points[j])).length()
+            j = i
+            i += 1
+        self.full_path_length = summ
+
+        # Load a model to follow path
+        self.model = self.art_obj_mgr.add_articulated_object_from_urdf(
+            filepath=self.user_metadata["urdf_path"], fixed_base=True
+        )
+        self.model.motion_type = phy.MotionType.KINEMATIC
+
+        # First update with step_size 0 to start character
+        self.update_pathfollower_sequential(step_size=0)
+
+    def update_pathfollower_sequential(
+        self, step_size: int = 1, path_time: Optional[float] = None
+    ):
+        """
+        [FILL AFTER IMPLEMENTATION]
+        """
+        # precondition
+        if not all(
+            [self.model, self.path_points, self.activity == Activity.PATH_FOLLOW_SEQ]
+        ):
+            return
+
+        walk = Move.walk_to_walk
+        global_neutral_correction = self.global_correction_quat(
+            mn.Vector3.z_axis(), mn.Vector3.x_axis()
+        )
+
+        # either take argument value for path_time or continue with cycle
+        self.path_time = path_time or (self.path_time + (step_size / 60.0))
+
+        # compute current frame from self.path_time w/ wrapping
+        mocap_time_length = walk.num_of_frames * (1.0 / 120.0)
+        mocap_cycles_past = math.floor(self.path_time / mocap_time_length)
+        mocap_time_curr = math.fmod(self.path_time, mocap_time_length)
+        mocap_frame = int(mocap_time_curr * 120)
+
+        # find distance progressed along shortest path
+        path_displacement = (
+            mocap_cycles_past * walk.map_of_total_displacement[LAST]
+            + walk.map_of_total_displacement[mocap_frame]
+        )
+
+        # handle wrapping or edgecase for dath displacement passing goal
+        if path_displacement > self.full_path_length:
+            self.path_time = 0.0
+            self.path_displacement = 0.0
```

Comment:
Should this be an fmod or other continuous wrapper instead of a hard reset to 0?

Reply:
My thinking is that when the character is expected to pass the full length of the path, it should reset to the origin of the path and begin again. This should not preserve any overflow of the path_displacement because the intended task is to move the character from point a to point b and the looping is merely a demo so if shouldn't need to exhibit wrapping. What do you think?

Reply:
Fair enough, this use case does focus on a one-way trip.